backend/src/composite/split.rs

```diff
@@ -0,0 +1,194 @@
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    ops::ControlFlow,
+    str::FromStr,
+};
+
+use powdr_ast::analyzed::SelectedExpressions;
+use powdr_ast::{
+    analyzed::{
+        AlgebraicExpression, Analyzed, Identity, IdentityKind, StatementIdentifier, Symbol,
+    },
+    parsed::{
+        asm::{AbsoluteSymbolPath, SymbolPath},
+        visitor::{ExpressionVisitable, VisitOrder},
+    },
+};
+use powdr_number::FieldElement;
+
+/// Splits a PIL into multiple PILs, one for each "machine".
+/// The rough algorithm is as follows:
+/// 1. The PIL is split into namespaces
+/// 2. Any lookups or permutations that reference multiple namespaces are removed.
+pub(crate) fn split_pil<F: FieldElement>(pil: Analyzed<F>) -> BTreeMap<String, Analyzed<F>> {
+    let statements_by_machine = split_by_namespace(&pil);
+
+    statements_by_machine
+        .into_iter()
+        .filter_map(|(machine_name, statements)| {
+            build_machine_pil(pil.clone(), statements).map(|pil| (machine_name, pil))
+        })
+        .collect()
+}
+
+/// Given a set of columns and a set of symbols, returns the columns that correspond to the symbols.
+pub(crate) fn select_machine_columns<'a, F: FieldElement>(
+    columns: &[(String, Vec<F>)],
+    symbols: impl Iterator<Item = &'a Symbol>,
+) -> Vec<(String, Vec<F>)> {
+    let names = symbols
+        .flat_map(|symbol| symbol.array_elements().map(|(name, _)| name))
+        .collect::<BTreeSet<_>>();
+    columns
+        .iter()
+        .filter(|(name, _)| names.contains(name))
+        .cloned()
+        .collect::<Vec<_>>()
+}
+
+/// From a symbol name, get the namespace of the symbol.
+fn get_namespace(name: &str) -> String {
```

Comment:
```suggestion
fn extract_namespace(name: &str) -> String {
```
a bit clearer maybe?